Draft implementation of plotted value formula

pjanik opened the pull request:
https://www.pivotaltracker.com/story/show/186267057

This is the first iteration of the plotted value formula implementation. It may appear a bit messy, and it's possible to break it while experimenting with it for sure. However, it should generally work as long as you always provide an attribute name, for example, `mean(Height)`, rather than just `mean()`, as seen in V2 (this will be another feature).

I've attempted to make the FormulaManager as generic as possible, implementing context-specific features in the form of adapters: `AttributeFormulaAdapter` and `PlottedValueFormulaAdapter`. These adapters need to implement a relatively simple interface and can utilize the API provided to them by the FormulaManager (trying to keep it minimal). I would appreciate any feedback on the pattern itself, naming, and the initialization or retrieval of things at runtime.

One aspect that requires further tweaking is the observation of dependencies. The FormulaManager establishes common observers, and each adapter can add custom ones (e.g., the plotted value observes the `setAttributeID` action, replicating what the GraphContentModel does). However, I believe that some of these common observers in the FormulaManager might need to be relocated to the attribute adapter. Additionally, the concept of 'casesToRecalculate' has become somewhat unclear, as it exists in attribute formulas only. Some renaming or clarification might be helpful.

MathJSScope has been adjusted to work with both attribute and plotted value formulas. I could potentially simplify it significantly for plotted value formulas and extend it for attribute formulas, but probably it's not a priority atm.

@emcelroy, I'm tagging you too because of the changes in plotted-value-adornment-model. I've overwritten some methods there to essentially skip the adornments update cycle and leave that to the FormulaManager and its adapters.

I'm opening this pull request now so that I can potentially address your feedback tomorrow before I'm out for a few days next week.

emcelroy commented on v3/src/models/data/plotted-value-formula-adapter.ts:
```diff
@@ -0,0 +1,146 @@
+import { getFormulaManager } from "../tiles/tile-environment"
+import { FormulaMathJsScope } from "./formula-mathjs-scope"
+
+// This should match type of PlottedValueAdornmentModel. Currently all these models use regular string instead of
+// enums / constants / literals, so I can only copy the value and save it here.
```
I should fix this. I made [a new chore story](https://www.pivotaltracker.com/story/show/186297171) for it.

emcelroy reviewed (approved):
This generally looks good to me 👍 The changes to the Plotted Value model seem reasonable and I double checked in the browser that it renders as expected in a variety of situations, and couldn't get it to break. So that's good.

I know you said this is a first iteration but just in case: there are `console.log` statements in attribute-formula-adapter.ts and plotted-value-formula-adapter.ts that don't seem necessary. But maybe you want to keep them for now?

kswenson commented on v3/src/models/data/plotted-value-formula-adapter.ts:
```diff
@@ -0,0 +1,146 @@
+  computeFormula(formulaContext: IFormulaContext, extraMetadata: IPlottedValueFormulaExtraMetadata,
+    childMostCases: ICase[]) {
+    const { formula, dataSet } = formulaContext
+    console.log(`[plotted value formula] recalculate "${formula.canonical}"`)
```
>> I know you said this is a first iteration but just in case: there are console.log statements in attribute-formula-adapter.ts and plotted-value-formula-adapter.ts that don't seem necessary. But maybe you want to keep them for now?
>
>These logs indicate when formulas are recalculated. If they are not too annoying for others, I'd like to keep them for now because it's probably the only way I can easily spot when a formula is recalculated when it shouldn't be. I remember @kswenson suggested using a debug toggle to show/hide them, but I'm worried that, in practice, no one will notice unwanted recalculations if we do that. So, I'll keep them in the near future if that's okay with everyone.

My inclination would be to wrap these in `if (DEBUG_FORMULAS)` and then if you want to have these logged by default for the time being we can default `DEBUG_FORMULAS` to `true` in `debug.ts`, but at least then there's only one place to change it rather than multiple places.

pjanik replied:
Okay, I've updated debug.ts and followed the existing pattern. So, console logs won't be visible by default now. I'll keep that debug option in my local storage.

kswenson reviewed (changes requested):
👍 Looks good and the general approach seems fine. Just a few relatively minor suggestions.

pjanik commented on v3/src/components/graph/models/graph-content-model.ts:
```diff
@@ -179,8 +188,19 @@ export const GraphContentModel = DataDisplayContentModel
           }
+
+          if (plottedValueFormulaAdapter) {
+            // TODO: This cast should not be necessary. Is it an MST quirk?
+            plottedValueFormulaAdapter.addGraphContentModel(self as IGraphContentModel)
```
Out of curiosity, is there some elegant solution to that?

pjanik commented on v3/src/models/data/formula-manager.ts:
```diff
@@ -1,48 +1,91 @@
 import { AddCasesAction, SetCaseValuesAction } from "./data-set-actions"
 import { IGlobalValueManager } from "../global/global-value-manager"
 import { IFormula } from "./formula"
+import { isAlive } from "@concord-consortium/mobx-state-tree"
```
Done in the new commits, thanks.

pjanik commented on v3/src/components/graph/models/graph-content-model.ts:
```diff
@@ -163,10 +170,12 @@ export const GraphContentModel = DataDisplayContentModel
-          return {sharedModelManager, sharedDataSets, tileSharedModels}
+          const plottedValueFormulaAdapter = getPlottedValueFormulaAdapter(self)
+
+          return {sharedModelManager, sharedDataSets, tileSharedModels, plottedValueFormulaAdapter}
```
Removed in the new commits, thanks.

pjanik commented on v3/src/components/graph/models/graph-content-model.ts:
```diff
@@ -82,6 +86,9 @@ export const GraphContentModel = DataDisplayContentModel
     get metadata() {
       return getTileCaseMetadata(self)
     },
+    get adornments(): IAdornmentModelUnion[] {
```
Done in the new commits, thanks.

pjanik commented on v3/src/models/data/plotted-value-formula-adapter.ts:
```diff
@@ -0,0 +1,146 @@
+  type = PLOTTED_VALUE_FORMULA_ADAPTER
+  api: IFormulaAdapterApi
+
+  @observable graphContentModels = new Map<string, IGraphContentModel>()
```
I've changed that to observable.shallow and did a similar update in the formula manager. Things seem to be working. Thanks!

pjanik commented:
Thank you for the feedback! I believe I have addressed all the suggestions, including hiding the console.log messages using the debug.ts helpers. The solutions were generally straightforward, so I will proceed with the merge without requesting another review to save your time. :)
